packages/wallet/ui/src/service/Dapps.js

```diff
@@ -0,0 +1,74 @@
+import { makeNotifierKit } from '@agoric/notifier';
+import {
+  loadDapps as load,
+  removeDapp as remove,
+  upsertDapp as upsert,
+} from '../store/Dapps.js';
+
+export const getDappService = publicAddress => {
+  const dapps = new Map();
+  const { notifier, updater } = makeNotifierKit();
+  const broadcastUpdates = () => updater.updateState([...dapps.values()]);
+
+  const upsertDapp = dapp => {
+    dapps.set(dapp.origin, dapp);
+    upsert(publicAddress, dapp);
+    broadcastUpdates();
+  };
+
+  const deleteDapp = origin => {
+    dapps.delete(origin);
+    remove(publicAddress, origin);
+    broadcastUpdates();
+  };
+
+  const setDappPetname = (origin, petname) => {
+    const dapp = dapps.get(origin);
+    if (!dapp) return;
+
+    const newValue = { ...dapp, petname };
+    upsertDapp(newValue);
+  };
+
+  const enableDapp = origin => {
+    const dapp = dapps.get(origin);
+    if (!dapp) return;
+
+    const newValue = { ...dapp, enable: true };
+    upsertDapp(newValue);
+  };
+
+  const storedDapps = load(publicAddress);
+  storedDapps.forEach(d => {
+    let enableAction;
+    const approvedP = new Promise(res => {
+      enableAction = () => {
+        enableDapp(d.origin);
+        res();
+      };
+    });
+    if (d.enable) {
+      enableAction();
```

Comment:
I have a hard time following the promise flow here. It looks like you want calling `enableAction` to `enableDapp` and then resolve `approvedP`. If `d.enable` is set then it should be called now and if it's not then let the dapp call it in `actions.enable`. I can't tell whether the dapp needs to call `enableAction` again if it's already succeeded (like does it still need to take `enableAction` if it's already enabled).

Though I tried to come up with something cleaner and I don't have anything better. Not requesting code changes but some comments.